include/openthread/platform/crypto.h

```diff
@@ -0,0 +1,441 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for PSA Crypto operations.
+ */
+
+#ifndef OPENTHREAD_PLATFORM_PSA_H_
+#define OPENTHREAD_PLATFORM_PSA_H_
+
+#include <stdint.h>
+
+#include "openthread-core-config.h"
+
+#include <mbedtls/aes.h>
+#include <mbedtls/md.h>
+#include <psa/crypto_types.h>
+#include <openthread/error.h>
+#include <openthread/platform/radio.h>
+#include "psa/crypto.h"
+#include "psa/crypto_values.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup plat-crypto
+ *
+ * @brief
+ *   This module includes the platform abstraction for Crypto.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This enumeration defines the crypto to use.
+ *
+ */
+typedef enum
+{
+    OT_CRYPTO_TYPE_MBEDTLS, ///< Use mbedTLS.
+    OT_CRYPTO_TYPE_PSA,     ///< Use ARM PSA.
+} otCryptoType;
+
+/**
+ * @struct otCryptoKey
+ *
+ * This structure represents the Key Material required for Crypto operations.
+ *
+ */
+
+OT_TOOL_PACKED_BEGIN
+struct otCryptoKey
+{
+    const uint8_t *mKey;       ///< Pointer to the buffer containing key. NULL indicates to use `mKeyRef`.
+    uint16_t       mKeyLength; ///< The key length in bytes (applicable when `mKey` is not NULL).
+    uint32_t       mKeyRef;    ///< The PSA key ref (requires `mKey` to be NULL).
+} OT_TOOL_PACKED_BEGIN;
+
+/**
+ *
+ * This structure represents the Key Material required for Crypto operations.
+ *
+ */
+typedef struct otCryptoKey otCryptoKey;
+
+/**
+ * Initialise the PSA module.
+ *
+ * @retval OT_ERROR_NONE          Successfully encrypted  @p aInput.
+ * @retval OT_ERROR_FAILED        Failed to encrypt @p aInput.
+ *
+ */
+otError otPlatCryptoInit(void);
+
+/**
+ * Import a key into PSA ITS.
+ *
+ * @param[out]  aKeyId            Reference to the key to be used for crypto operations.
+ * @param[in]   aKeyType          Key Type encoding for the key.
+ * @param[in]   aKeyAlgorithm     Key algorithm encoding for the key.
+ * @param[in]   aKeyUsage         Key Usage encoding for the key.
+ * @param[in]   aKeyPersistence   Key Persistence for this key
+ * @param[in]   aKey              Actual key to be imported.
+ * @param[in]   aKeyLen           Length of the key to be imported.
+ *
+ * @retval OT_ERROR_NONE          Successfully encrypted  @p aInput.
+ * @retval OT_ERROR_FAILED        Failed to encrypt @p aInput.
+ * @retval OT_ERROR_INVALID_ARGS  @p aInput or @p aOutput was set to NULL.
+ *
+ */
+otError otPlatCryptoImportKey(psa_key_id_t *        aKeyId,
+                              psa_key_type_t        aKeyType,
+                              psa_algorithm_t       aKeyAlgorithm,
+                              psa_key_usage_t       aKeyUsage,
+                              psa_key_persistence_t aKeyPersistence,
```

Comment:
I think it would be better to not assume implementation in platform abstraction interface. Using `psa_` types here looks like assumption that implementation of this function is PSA. If we want to support this crypto abstraction, I think these parameters should be also abstracted to OT-defined types, not PSA-defined types.

Reply:
Internally all the `psa_` types are basic types like uint32. I can typedef these to `OtCrypto` types. But that still will be an assumption specific to PSA. Is there better way we can let the vendors define this? Maybe add a crypto_config header vendors can use?

Reply:
I think the correct way would be to define types like: `typedef uint32_t otKeyType` and its values `#define OT_KEY_TYPE_AES 0x01`. It should include only parameters and values used by OpenThread (not all defined in mbedtls, or PSA, or other implementation).

In PSA implementation of this API there should be a map from OT-defined values and types to PSA-defined like:
```
static psa_key_type_t get_key_type(otKeyType type)
{
  switch (type) {
    case OT_KEY_TYPE_AES:
      return PSA_KEY_TYPE_AES;
}
```

I think this would be the cleanest way to define such abstraction.

Reply:
@hubertmis I have completed the changes we discussed above. Can you have a look.